src/maxminddb/lib.rs

```diff
@@ -382,25 +390,131 @@ impl BinaryDecoder {
     }
 }
 
-/// A reader for the MaxMind DB format
-pub struct Reader {
-    decoder: BinaryDecoder,
+/// A reader for the MaxMind DB format using `mmap` internally to only cause a minimal set of IO operations to happen, instead of reading the entire database (potentially 100s of MiB) into memory.
+#[cfg(feature = "mmap")]
+#[allow(dead_code)]
+pub struct OwnedReader<'data> {
+    data: Mmap,
+    inner: Reader<'data>,
+}
+
+/// A type definition substituting `OwnedReader<'data>` with `OwnedReaderFile<'data>`, used if the feature `mmap` is not active.
+#[cfg(not(feature = "mmap"))]
+pub type OwnedReader<'data> = OwnedReaderFile<'data>;
+
+/// A reader for the MaxMind DB format using a `Vec<u8>` internally to hold a complete copy of the database.
+#[allow(dead_code)]
+pub struct OwnedReaderFile<'data> {
+    data: Vec<u8>,
+    inner: Reader<'data>,
+}
+
+/// Allows to use OwnedReader as a drop-in replacement for a `Reader<'data>`.
+impl<'data> Deref for OwnedReader<'data> {
+    type Target = Reader<'data>;
+
+    fn deref(&self) -> &<Self as Deref>::Target {
+        &self.inner
+    }
+}
+
+/// A reader for the MaxMind DB format. The lifetime `'data` is tied to the lifetime of the underlying buffer holding the contents of the database file.
+pub struct Reader<'data> {
+    decoder: BinaryDecoder<'data>,
     pub metadata: Metadata,
     ipv4_start: usize,
 }
 
-impl<'de> Reader {
+#[cfg(feature = "mmap")]
+impl<'de, 'data> Reader<'data> {
     /// Open a MaxMind DB database file.
     ///
     /// # Example
     ///
     /// ```
     /// let reader = maxminddb::Reader::open("test-data/test-data/GeoIP2-City-Test.mmdb").unwrap();
     /// ```
-    pub fn open(database: &str) -> Result<Reader, MaxMindDBError> {
+    pub fn open(database: &str) -> Result<OwnedReader<'static>, MaxMindDBError> {
+        Self::open_mmap(database)
+    }
+
+    pub fn open_mmap(database: &str) -> Result<OwnedReader<'static>, MaxMindDBError> {
```

Comment:
I know you are are just copying above, but I think `AsRef<Path>` would be more idiomatic. I have been meaning to switch to that generally.